File: 2020/22/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	input := strings.Split(string(data), "\n")
	parsedInput := parseInput(input)
	fmt.Println(part1(parsedInput))
	fmt.Println(part2(parsedInput))
}

type Player struct {
	deck []int
}

func (p *Player) draw() int {
	card := p.deck[0]
	p.deck = p.deck[1:]
	return card
}

func (p *Player) add(cards ...int) {
	p.deck = append(p.deck, cards...)
}

func (p Player) score() int {
	score := 0
	for i, card := range p.deck {
		score += card * (len(p.deck) - i)
	}
	return score
}

func parseInput(input []string) []Player {
	players := []Player{}
	player1 := Player{}
	player2 := Player{}
	row := 1
	for len(input[row]) > 0 {
		card, _ := strconv.Atoi(input[row])
		player1.add(card)
		row++
	}
	row += 2
	for row < len(input) && len(input[row]) > 0 {
		card, _ := strconv.Atoi(input[row])
		player2.add(card)
		row++
	}
	players = append(players, player1, player2)
	return players
}

func part1(players []Player) int {
	player1 := players[0]
	player2 := players[1]
	for len(player1.deck) > 0 && len(player2.deck) > 0 {
		card1 := player1.draw()
		card2 := player2.draw()
		if card1 > card2 {
			player1.add(card1, card2)
		} else {
			player2.add(card2, card1)
		}
	}
	if len(player1.deck) > 0 {
		return player1.score()
	}
	return player2.score()
}

func recordHistory(player1 Player, player2 Player) string {
	history1 := []string{}
	history2 := []string{}
	for _, card := range player1.deck {
		history1 = append(history1, strconv.Itoa(card))
	}
	for _, card := range player2.deck {
		history2 = append(history2, strconv.Itoa(card))
	}
	return strings.Join(history1, ",") + "|" + strings.Join(history2, ",")
}

func recurCombat(player1 Player, player2 Player) (int, int) {
	history := map[string]bool{}
	for len(player1.deck) > 0 && len(player2.deck) > 0 {
		currHistory := recordHistory(player1, player2)
		if history[currHistory] {
			return 1, player1.score()
		}
		history[currHistory] = true
		card1 := player1.draw()
		card2 := player2.draw()
		if len(player1.deck) >= card1 && len(player2.deck) >= card2 {
			subPlayer1 := Player{}
			subPlayer2 := Player{}
			subdeck1 := make([]int, card1)
			subdeck2 := make([]int, card2)
			copy(subdeck1, player1.deck)
			copy(subdeck2, player2.deck)
			subPlayer1.deck = subdeck1
			subPlayer2.deck = subdeck2
			winner, _ := recurCombat(subPlayer1, subPlayer2)
			if winner == 1 {
				player1.add(card1, card2)
			} else {
				player2.add(card2, card1)
			}
		} else if card1 > card2 {
			player1.add(card1, card2)
		} else {
			player2.add(card2, card1)
		}
	}
	if len(player1.deck) > 0 {
		return 1, player1.score()
	}
	return 2, player2.score()
}

func part2(players []Player) int {
	player1 := players[0]
	player2 := players[1]
	_, score := recurCombat(player1, player2)
	return score
}

File: 2020/22/solution.ts
import * as fs from "fs";
import * as path from "path";

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n\n");

class Player {
  deck: number[];
  constructor(deck: number[]) {
    this.deck = deck;
  }

  draw(): number {
    if (this.deck.length === 0) return -1;
    return this.deck.shift()!;
  }

  play(player2: Player): number {
    const card1 = this.draw();
    const card2 = player2.draw();
    if (card1 > card2) {
      this.deck.push(card1, card2);
      if (player2.deck.length === 0) return 1;
    } else {
      player2.deck.push(card2, card1);
      if (this.deck.length === 0) return 2;
    }
    return 0;
  }

  get score(): number {
    return this.deck.reduce((a, b, i) => a + b * (this.deck.length - i), 0);
  }

  copy(n: number): Player {
    return new Player(this.deck.slice(0, n));
  }
}

const player1 = new Player(input[0].split("\n").slice(1).map(Number));
const player2 = new Player(input[1].split("\n").slice(1).map(Number));

let winner = 0;
while ((winner = player1.play(player2)) === 0) {}
if (winner === 1) console.log(player1.score);
else console.log(player2.score);

function historyKey(player1: Player, player2: Player): string {
  return player1.deck.join(",") + "|" + player2.deck.join(",");
}

function recursiveCombat(player1: Player, player2: Player): number {
  const combatHistory: Set<string> = new Set();
  while (true) {
    const currKey = historyKey(player1, player2);
    if (combatHistory.has(currKey)) {
      return 1;
    }
    combatHistory.add(currKey);
    const card1 = player1.draw();
    const card2 = player2.draw();
    if (player1.deck.length >= card1 && player2.deck.length >= card2) {
      const winner = recursiveCombat(player1.copy(card1), player2.copy(card2));
      if (winner === 1) player1.deck.push(card1, card2);
      else player2.deck.push(card2, card1);
    } else {
      if (card1 > card2) player1.deck.push(card1, card2);
      else player2.deck.push(card2, card1);
    }
    if (player1.deck.length === 0) return 2;
    if (player2.deck.length === 0) return 1;
  }
}

const player1Recur = new Player(input[0].split("\n").slice(1).map(Number));
const player2Recur = new Player(input[1].split("\n").slice(1).map(Number));

winner = recursiveCombat(player1Recur, player2Recur);
if (winner === 1) console.log(player1Recur.score);
else console.log(player2Recur.score);
